File: templates/monitoring/prometheus_middleware.go
package monitoring

import (
	"net/http"
	"strconv"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	httpRequestsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "app_http_requests_total",
			Help: "Total HTTP requests",
		},
		[]string{"method", "path", "status"},
	)

	httpRequestDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "app_http_request_duration_seconds",
			Help:    "HTTP request latency",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"method", "path"},
	)

	httpRequestsInProgress = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "app_http_requests_in_progress",
			Help: "HTTP requests in progress",
		},
		[]string{"method", "path"},
	)
)

func init() {
	// Register metrics
	prometheus.MustRegister(httpRequestsTotal)
	prometheus.MustRegister(httpRequestDuration)
	prometheus.MustRegister(httpRequestsInProgress)
}

// PrometheusMiddleware wraps HTTP handlers with Prometheus metrics
func PrometheusMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		path := r.URL.Path
		method := r.Method

		// Track in-progress requests
		httpRequestsInProgress.WithLabelValues(method, path).Inc()
		defer httpRequestsInProgress.WithLabelValues(method, path).Dec()

		// Wrap response writer to capture status code
		wrapped := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}

		// Call next handler
		next.ServeHTTP(wrapped, r)

		// Record metrics
		duration := time.Since(start).Seconds()
		status := strconv.Itoa(wrapped.statusCode)

		httpRequestsTotal.WithLabelValues(method, path, status).Inc()
		httpRequestDuration.WithLabelValues(method, path).Observe(duration)
	})
}

// responseWriter wraps http.ResponseWriter to capture status code
type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// MetricsHandler returns the Prometheus metrics handler
func MetricsHandler() http.Handler {
	return promhttp.Handler()
}

// Example usage:
// 
// func main() {
//     mux := http.NewServeMux()
//     
//     // Your routes
//     mux.HandleFunc("/api/users", handleUsers)
//     
//     // Metrics endpoint
//     mux.Handle("/metrics", monitoring.MetricsHandler())
//     
//     // Wrap with Prometheus middleware
//     handler := monitoring.PrometheusMiddleware(mux)
//     
//     http.ListenAndServe(":8000", handler)
// }

File: templates/monitoring/prometheus_middleware.js
/**
 * Prometheus metrics middleware for Node.js/Express
 * Auto-tracks HTTP requests with path, method, status code
 */
const promClient = require('prom-client');

// Create metrics
const httpRequestsTotal = new promClient.Counter({
  name: 'app_http_requests_total',
  help: 'Total HTTP requests',
  labelNames: ['method', 'path', 'status']
});

const httpRequestDuration = new promClient.Histogram({
  name: 'app_http_request_duration_seconds',
  help: 'HTTP request latency',
  labelNames: ['method', 'path'],
  buckets: [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5]
});

const httpRequestsInProgress = new promClient.Gauge({
  name: 'app_http_requests_in_progress',
  help: 'HTTP requests in progress',
  labelNames: ['method', 'path']
});

// Enable default metrics (CPU, memory, etc.)
promClient.collectDefaultMetrics();

/**
 * Express middleware for Prometheus metrics
 */
function prometheusMiddleware(req, res, next) {
  const start = Date.now();
  const path = req.route ? req.route.path : req.path;
  const method = req.method;
  
  // Track in-progress requests
  httpRequestsInProgress.labels(method, path).inc();
  
  // Capture response
  res.on('finish', () => {
    const duration = (Date.now() - start) / 1000;
    const status = res.statusCode;
    
    // Record metrics
    httpRequestsTotal.labels(method, path, status).inc();
    httpRequestDuration.labels(method, path).observe(duration);
    httpRequestsInProgress.labels(method, path).dec();
  });
  
  next();
}

/**
 * Metrics endpoint handler
 */
async function metricsHandler(req, res) {
  res.set('Content-Type', promClient.register.contentType);
  res.end(await promClient.register.metrics());
}

/**
 * Setup Prometheus metrics for Express app
 */
function setupExpressMetrics(app) {
  // Add middleware
  app.use(prometheusMiddleware);
  
  // Add metrics endpoint
  app.get('/metrics', metricsHandler);
  
  return app;
}

module.exports = {
  prometheusMiddleware,
  metricsHandler,
  setupExpressMetrics,
  httpRequestsTotal,
  httpRequestDuration,
  httpRequestsInProgress
};
